include/proxy-wasm/context.h

```diff
@@ -0,0 +1,306 @@
+
+  /**
+   * Calls into the VM.
+   * These are implemented by the proxy-independent host code. They are virtual to support some
+   * types of testing.
+   */
+
+  // Context
+  void onCreate(uint32_t parent_context_id) override;
+  bool onDone() override;
+  void onDelete() override;
+
+  // Root
+  bool onStart(std::shared_ptr<PluginBase> plugin) override;
+  bool onConfigure(std::shared_ptr<PluginBase> plugin) override;
+  void onHttpCallResponse(HttpCallToken token) override;
+  void onQueueReady(uint32_t SharedQueueDequeueToken) override;
+  void onGrpcReceiveInitialMetadata(GrpcToken token) override;
+  void onGrpcReceive(GrpcToken token) override;
+  void onGrpcReceiveTrailingMetadata(GrpcToken token) override;
+  void onGrpcClose(GrpcToken token, const GrpcStatus &status, const string_view message) override;
+
+  // Stream (e.g. HTTP, Network)
+  void onFinalized() override;
+  const BufferInterface *getBuffer(WasmBufferType /* type */) override {
+    unimplemented();
+    return nullptr;
+  }
+  bool end_of_stream(WasmStreamType /* type */) override {
+    unimplemented();
+    return true;
+  }
+  WasmResult continueStream(WasmStreamType /* type */) override { return unimplemented(); }
+
+  // HTTP
+  ProxyAction onRequestHeaders() override;
+  ProxyAction onRequestBody() override;
+  ProxyAction onRequestTrailers() override;
+  ProxyAction onRequestMetadata() override;
+  ProxyAction onResponseHeaders() override;
+  ProxyAction onResponseBody() override;
+  ProxyAction onResponseTrailers() override;
+  ProxyAction onResponseMetadata() override;
+  WasmResult sendLocalResponse(uint64_t /* response_code */, string_view /* body */,
+                               Pairs /* additional_headers */, uint64_t /* grpc_status */,
+                               string_view /* details */) override {
+    return unimplemented();
+  }
+
+  // Network
+  ProxyAction onNetworkNewConnection() override;
+  ProxyAction onDownstreamData() override;
+  ProxyAction onUpstreamData() override;
+  using CloseType = NetworkInterface::CloseType;
+  void onDownstreamConnectionClose(CloseType close_type) override;
+  void onUpstreamConnectionClose(CloseType close_type) override;
+
+  // General
+  void error(string_view message) override {
+    std::cerr << message << "\n";
+    abort();
+  }
+  WasmResult unimplemented() override {
+    error("unimplemented proxy-wasm API");
+    return WasmResult::Unimplemented;
+  }
+  WasmResult log(uint64_t level, string_view message) override {
+    if (level >= static_cast<uint32_t>(LogLevel::error)) {
+      std::cerr << log_prefix() << message << "\n";
+    } else {
+      std::cout << log_prefix() << message << "\n";
+    }
+    return WasmResult::Ok;
+  }
+  virtual WasmResult setTimerPeriod(std::chrono::milliseconds /* period */,
+                                    uint32_t * /* timer_token_ptr */) override {
+    return unimplemented();
+  }
+  uint64_t getCurrentTimeNanoseconds() override {
```

Comment:
use TimeSystem::systemTime() for this. Then you can drop the include of time.h. In any case I think this is not a good candidate for inlining.

Reply:
That is the envoy-specific implementation.  See the companion PR https://github.com/envoyproxy/envoy/pull/10262 for details.  We use a TimeSource for mock-ability.

Reply:
can we relegate this to a cc file? Seems like we don't need to expose this as a header; might also not be windows-compatible.